### PR TITLE
Use MetricType to specify the metric / distance compute function

### PR DESCRIPTION
**Pull request by eddyxu**
* Persist MetricType in index protobuf
* `lq index --metric-type [l2|cosine]`

### CODE REVIEW

**Comment by eddyxu** [comment]
Will add the support via pyo3 in a following PR.

File: rust/src/utils/distance/cosine.rs [hunk 1]
```diff
@@ -101,41 +94,41 @@ fn cosine_dist_simd(from: &Float32Array, to: &Float32Array, dimension: usize) ->
     for y in to_values.chunks_exact(dimension) {
         #[cfg(any(target_arch = "aarch64"))]
         {
-            builder.append_value(unsafe { cosine_dist_neon(x, y, x_norm) });
+            builder.append_value(unsafe { 1.0 - cosine_dist_neon(x, y, x_norm) });
```
Comment (changhiskhan) [inline rust/src/utils/distance/cosine.rs:97]: This should happen inside the distance function OR the fn should be renamed cosine_similarity_neon/fma

**Review by changhiskhan — COMMENTED** [review]
Minor questions

File: rust/src/dataset/scanner.rs [hunk 2]
```diff
@@ -199,6 +200,14 @@ impl Scanner {
         self
     }
 
+    /// Change the distance [MetricType], i.e, L2 or Cosine distance.
+    pub fn distance_metrics(&mut self, metric_type: MetricType) -> &mut Self {
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply rust/src/dataset/scanner.rs:204]: fixed.

File: rust/src/dataset.rs [hunk 3]
```diff
@@ -388,6 +388,7 @@ impl Dataset {
                     column,
                     vec_params.num_partitions,
                     vec_params.num_sub_vectors,
+                    vec_params.metrics_type.clone(),
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply rust/src/dataset.rs:391]: fixed.

File: rust/src/utils/distance/cosine.rs (hunk 1 above)
Comment (eddyxu) [reply rust/src/utils/distance/cosine.rs]: fixed.

**Review by changhiskhan — APPROVED** [review]

**Pull request merged**